cpp/include/raft/cluster/detail/kmeans.cuh

```diff
@@ -19,967 +19,1056 @@
-static __global__ void minDistances2(index_type_t n,
-                                     value_type_t* __restrict__ dists_old,
-                                     const value_type_t* __restrict__ dists_new,
-                                     index_type_t* __restrict__ codes_old,
-                                     index_type_t code_new)
+template <typename DataT, typename IndexT>
+void initKMeansPlusPlus(const raft::handle_t& handle,
+                        const KMeansParams& params,
+                        raft::device_matrix_view<const DataT> X,
+                        rmm::device_uvector<DataT>& centroidsRawData,
+                        rmm::device_uvector<char>& workspace)
 {
-  // Loop index
-  index_type_t i = threadIdx.x + blockIdx.x * blockDim.x;
-
-  // Distances
-  value_type_t dist_old_private;
-  value_type_t dist_new_private;
-
-  // Each row is processed by a thread
-  while (i < n) {
-    // Get old and new distances
-    dist_old_private = dists_old[i];
-    dist_new_private = dists_new[i];
-
-    // Update if new distance is smaller than old distance
-    if (dist_new_private < dist_old_private) {
-      dists_old[i] = dist_new_private;
-      codes_old[i] = code_new;
-    }
-
-    // Move to another row
-    i += blockDim.x * gridDim.x;
-  }
+  cudaStream_t stream = handle.get_stream();
+  auto n_samples      = X.extent(0);
+  auto n_features     = X.extent(1);
+  auto n_clusters     = params.n_clusters;
+  auto metric         = static_cast<raft::distance::DistanceType>(params.metric);
+  centroidsRawData.resize(n_clusters * n_features, stream);
+  kmeansPlusPlus<DataT, IndexT>(handle, params, X, metric, workspace, centroidsRawData, stream);
 }
 
-/**
- *  @brief Compute size of k-means clusters.
- *    Block and grid dimensions should be 1-dimensional. Ideally the
- *    grid is large enough so there are n threads.
- *  @tparam index_type_t the type of data used for indexing.
- *  @param n Number of observation vectors.
- *  @param k Number of clusters.
- *  @param codes (Input, n entries) Cluster assignments.
- *  @param clusterSizes (Output, k entries) Number of points in each
- *    cluster. Entries must be initialized to zero.
- */
-template <typename index_type_t>
-static __global__ void computeClusterSizes(index_type_t n,
-                                           const index_type_t* __restrict__ codes,
-                                           index_type_t* __restrict__ clusterSizes)
+template <typename DataT, typename IndexT>
+void kmeans_fit_main(const raft::handle_t& handle,
+                     const KMeansParams& params,
+                     raft::device_matrix_view<const DataT> X,
+                     raft::device_vector_view<const DataT> weight,
+                     rmm::device_uvector<DataT>& centroidsRawData,
+                     DataT& inertia,
+                     IndexT& n_iter,
+                     rmm::device_uvector<char>& workspace)
```

Comment:
The workspace can't be a device_vector_view because it needs to be resized inside the function.

Reply:
Can you add include guards in core/logger.hpp for now? I think this is happening because both the deprecated and new header files are being included somewhere.

Reply:
I added the include guards but it didn't solve it. Adding inline to the 3 functions being define multiple times solved it